api/src/main/java/io/druid/data/input/impl/AbstractTextFilesFirehoseFactory.java

```diff
@@ -0,0 +1,151 @@
+    implements FirehoseFactory<StringInputRowParser>
+{
+  private static final Logger LOG = new Logger(AbstractTextFilesFirehoseFactory.class);
+
+  private final List<ObjectType> objects;
+
+  public AbstractTextFilesFirehoseFactory(Collection<ObjectType> objects)
+  {
+    this.objects = ImmutableList.copyOf(Preconditions.checkNotNull(objects));
+  }
+
+  @Override
+  public Firehose connect(StringInputRowParser firehoseParser) throws IOException
+  {
+    final Iterator<ObjectType> iterator = objects.iterator();
+    return new FileIteratingFirehose(
+        new Iterator<LineIterator>()
+        {
+          @Override
+          public boolean hasNext()
+          {
+            return iterator.hasNext();
+          }
+
+          @Override
+          public LineIterator next()
+          {
+            if (!hasNext()) {
+              throw new NoSuchElementException();
+            }
+            final ObjectType object = iterator.next();
+            try {
+              final InputStream stream = openStream(object);
+              return IOUtils.lineIterator(
+                  new BufferedReader(
+                      new InputStreamReader(wrapIfNeeded(object, stream), Charsets.UTF_8)
+                  )
+              );
+            }
+            catch (Exception e) {
+              LOG.error(
+                  e,
+                  "Exception reading object[%s]",
+                  object
+              );
+              throw Throwables.propagate(e);
+            }
+          }
+        },
+        firehoseParser
+    );
+  }
+
+  public List<ObjectType> getObjects()
+  {
+    return objects;
+  }
+
+  /**
+   * Wrap the given stream if needed, currently when the given object is compressed with gzip.
+   *
+   * @param object an object
+   * @param innerStream input stream for object
+   *
+   * @return wrapped input stream if the object is gzipped
+   *
+   * @throws IOException
+   */
+  protected InputStream wrapIfNeeded(ObjectType object, InputStream innerStream) throws IOException
+  {
+    return wrapIfNeeded(innerStream, isGzipped(object));
+  }
+
+  protected static InputStream wrapIfNeeded(InputStream innerStream, boolean isGzipped) throws IOException
+  {
+    if (isGzipped) {
+      return CompressionUtils.gzipInputStream(innerStream);
+    }
+
+    return innerStream;
+  }
+
+  /**
+   * Open an input stream from the given object.
+   * The result input stream must not be wrapped even if the object is compressed with gzip.
+   *
+   * @param object an object to be read
+   *
+   * @return an input stream for the object
+   *
+   * @throws IOException
+   */
+  protected abstract InputStream openStream(ObjectType object) throws IOException;
+
+  /**
+   * Check if the object is compressed with gzip.
+   *
+   * @param object an object to be checked
+   *
+   * @return true if the object is compressed with gzip
+   */
+  protected abstract boolean isGzipped(ObjectType object);
```

Comment:
Is it necessary to have isGzipped and wrapIfNeeded? I would think that the implementations could wrap if needed, since they should know if an object is gzipped or not.